### PR TITLE
code improvements, readability enhancements, type safety, modern syntax adaptation

### PR DESCRIPTION
**Pull request by xHeaven**
See commit messages. As far as I know, there are no breaking changes.

### CODE REVIEW

File: lib/Providers/Qr/QRicketProvider.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ public function getQRCodeImage(string $qrtext, int $size): string
     public function getUrl(string $qrtext, int $size): string
     {
         return 'http://qrickit.com/api/qr'
-            . '?qrsize=' . (string) $size
+            . '?qrsize=' . $size
```
Comment (NicolasCARPi) [inline lib/Providers/Qr/QRicketProvider.php:38]: why did you remove the cast to string? `size` is `int` here, and in a concat, one needs to cast to string first. Anyway, using `sprintf` would be better, no?

File: lib/Providers/Time/HttpTimeProvider.php [hunk 2]
```diff
@@ -21,10 +21,8 @@ class HttpTimeProvider implements ITimeProvider
     public function __construct(
         public string $url = 'https://google.com',
         public string $expectedtimeformat = 'D, d M Y H:i:s O+',
-        array $options = null,
+        ?array        $options = null,
```
Comment (NicolasCARPi) [inline lib/Providers/Time/HttpTimeProvider.php:24]: I'm afraid this kind of alignment spaces will be removed by an automatic formatter such as phpcs

**Review by NicolasCARPi — COMMENTED** [review]
So I went over the code changes. This is not a refactor like the title implies, but rather a set of changes to make the code use modern syntax. Use match instead of switch, use null coalescing operator, etc... I didn't test it but it should indeed not have any impact on code behavior.

File: lib/Providers/Qr/QRicketProvider.php (hunk 1 above)
Comment (xHeaven) [reply lib/Providers/Qr/QRicketProvider.php:38]: PHP has type juggling. If you concatenate anything with a string that is not a string type, PHP will automatically convert it. The type cast is absolutely unnecessary there. See: https://onlinephp.io/c/e7113
Eg.
```php
$hello = "Hello";
$hello123 = $hello . 123; // Hello123
```

File: lib/Providers/Time/HttpTimeProvider.php (hunk 2 above)
Comment (xHeaven) [reply lib/Providers/Time/HttpTimeProvider.php:24]: It was PHP-CS-Fixer that put it there in the first place, I've used it as my external formatter. The code style and the code altogether in the PR are entirely validated with php-cs-fixer, phpstan, and unit tests too.

**Comment by xHeaven** [comment]
> So I went over the code changes. This is not a refactor like the title implies, but rather a set of changes to make the code use modern syntax. Use match instead of switch, use null coalescing operator, etc... I didn't test it but it should indeed not have any impact on code behavior.

As far as I'm aware, refactoring means introducing changes in an existing codebase in a way that doesn't change the software's behavior but improves some aspects of it, such as readability, performance, etc.
If you've got a different definition, I'm all ears, I'm also open to changing the title if you've got something more descriptive. Let me know.

File: lib/Providers/Qr/QRicketProvider.php (hunk 1 above)
Comment (NicolasCARPi) [reply lib/Providers/Qr/QRicketProvider.php:38]: Don't get me wrong, I'm not saying it is *necessary*, but rather *improper*, and the whole point is to avoid as much as possible type juggling and make things as explicit as possible. And I know of at least one static analysis tool (maybe rector?) that gives an error when concat str/int is encountered. So yeah, use sprintf!

File: lib/Providers/Time/HttpTimeProvider.php (hunk 2 above)
Comment (NicolasCARPi) [reply lib/Providers/Time/HttpTimeProvider.php:24]: very well then! (I was the one adding the .phpcs-fixer config file ^^)

**Comment by NicolasCARPi** [comment]
> As far as I'm aware, refactoring means introducing changes in an existing codebase in a way that doesn't change the software's behavior but improves some aspects of it, such as readability, performance, etc. If you've got a different definition

To me, refactoring means having an impact on the private API, and maybe public, by splitting/merging classes, and mainly doing _refactoring_ of functions, in the algebraic/mathematical sense of the term.

But honestly your definition works too and nobody really care. It's just that reading "massive refactoring" I was a bit scared, but in fact it's more like "many small code improvements with no effective changes" ;)

File: lib/Providers/Qr/QRicketProvider.php (hunk 1 above)
Comment (xHeaven) [reply lib/Providers/Qr/QRicketProvider.php]: I've just updated the `getUrl` implementations to use `sprintf` in https://github.com/RobThree/TwoFactorAuth/pull/108/commits/4d91d6ae27e3b23d0af3be550956b8248a57ac5b.
Let me know your thoughts!

Comment (NicolasCARPi) [reply lib/Providers/Qr/QRicketProvider.php]: I think we lose a lot on readability here. You can leave it like that it's alright, it's not a big deal, but I'm wondering if using `http_build_query` from an array wouldn't be better in term of readablity of the code :thinking: As a bonus, we would have proper url encoding!

Comment (xHeaven) [reply lib/Providers/Qr/QRicketProvider.php]: Check https://github.com/RobThree/TwoFactorAuth/pull/108/commits/cf84371dad11591b77b14c5dea5ac7c52a288b22, I think it improves a lot in terms of readability. Could it be improved further?

Comment (NicolasCARPi) [reply lib/Providers/Qr/QRicketProvider.php]: Oh yeah, definitely much better and it allows to remove the urlencode stuff! :+1:

File: lib/Providers/Qr/BaconQrCodeProvider.php [hunk 3]
```diff
@@ -23,10 +22,10 @@ class BaconQrCodeProvider implements IQRCodeProvider
      * Ensure we using the latest Bacon QR Code and specify default options
      */
     public function __construct(
-        private int $borderWidth = 4,
+        private readonly int $borderWidth = 4,
```
Comment (willpower232) [inline lib/Providers/Qr/BaconQrCodeProvider.php:25]: I don't think there is much point in making private properties readonly as they can only be used within this class so it feels like a waste of effort.

Also why aren't the other properties defined as readonly? What makes this one deserving of readonly and not the others?

File: lib/Providers/Qr/BaconQrCodeProvider.php [hunk 4]
```diff
@@ -23,10 +22,10 @@ class BaconQrCodeProvider implements IQRCodeProvider
...
         private string|array $backgroundColour = '#ffffff',
         private string|array $foregroundColour = '#000000',
-        private string $format = 'png',
+        private string       $format = 'png',
```
Comment (willpower232) [inline lib/Providers/Qr/BaconQrCodeProvider.php:28]: I can appreciate the nice visuals of having all the variables line up but the practice in general usually ends up with needing to micro-manage spaces on unrelated lines when making changes so I'd say this wasn't worth it

File: lib/Providers/Qr/BaconQrCodeProvider.php [hunk 5]
```diff
@@ -52,22 +51,17 @@ public function getMimeType(): string
         throw new RuntimeException(sprintf('Unknown MIME-type: %s', $this->format));
     }
 
-    public function getQRCodeImage(string $qrText, int $size): string
+    public function getQRCodeImage(string $qrtext, int $size): string
```
Comment (willpower232) [inline lib/Providers/Qr/BaconQrCodeProvider.php:54]: camelCase should be used here

File: lib/Providers/Qr/BaconQrCodeProvider.php [hunk 6]
```diff
@@ -84,7 +78,7 @@ private function getQRCodeByBackend($qrText, $size, ImageBackEndInterface $backe
         $rendererStyleArgs = array($size, $this->borderWidth);
 
         if (is_array($this->foregroundColour) && is_array($this->backgroundColour)) {
-            $rendererStyleArgs = array_merge($rendererStyleArgs, array(
+            $rendererStyleArgs = array(...$rendererStyleArgs, ...array(
```
Comment (willpower232) [inline lib/Providers/Qr/BaconQrCodeProvider.php:81]: whilst this change is indeed equivalent, I would argue that it is less obvious what is happening now, array_merge is more readable in my opinion. I would say though that replacing `array()` with `[]` would be good to see (i.e. `array_merge($rendererStyleArgs, [`)

File: lib/Providers/Time/HttpTimeProvider.php (hunk 2 above)
Comment (willpower232) [inline lib/Providers/Time/HttpTimeProvider.php:24]: missing explicit public/private and potentially annoying spaces as previously mentioned

File: lib/Providers/Time/NTPTimeProvider.php [hunk 7]
```diff
@@ -47,7 +47,7 @@ public function getTime()
 
             // Interpret response
             $data = unpack('N12', $recv);
-            $timestamp = (int) sprintf('%u', $data[9]);
+            $timestamp = (int)sprintf('%u', $data[9]);
```
Comment (willpower232) [inline lib/Providers/Time/NTPTimeProvider.php:50]: I find a space after the type coercion much more readable

File: lib/TwoFactorAuth.php [hunk 8]
```diff
@@ -28,13 +28,13 @@ class TwoFactorAuth
...
-        private ?IQRCodeProvider $qrcodeprovider = null,
-        private ?IRNGProvider $rngprovider = null,
-        private ?ITimeProvider $timeprovider = null
+        private readonly ?string   $issuer = null,
+        private readonly int       $digits = 6,
+        private readonly int       $period = 30,
+        private readonly Algorithm $algorithm = Algorithm::Sha1,
+        private ?IQRCodeProvider   $qrcodeprovider = null,
+        private ?IRNGProvider      $rngprovider = null,
+        private ?ITimeProvider     $timeprovider = null
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:37]: potentially unnecessary readonly and potentially annoying spaces as previously discussed

File: lib/TwoFactorAuth.php [hunk 9]
```diff
@@ -54,7 +54,7 @@ public function __construct(
     public function createSecret(int $bits = 80, bool $requirecryptosecure = true): string
     {
         $secret = '';
-        $bytes = (int) ceil($bits / 5);   // We use 5 bits of each byte (since we have a 32-character 'alphabet' / BASE32)
+        $bytes = (int)ceil($bits / 5);   // We use 5 bits of each byte (since we have a 32-character 'alphabet' / BASE32)
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:57]: I find a space after the type coercion much more readable

File: lib/TwoFactorAuth.php [hunk 10]
```diff
@@ -79,7 +79,7 @@ public function getCode(string $secret, ?int $time = null): string
         $value = unpack('N', $hashpart);                                                   // Unpack binary value
         $value = $value[1] & 0x7FFFFFFF;                                                   // Drop MSB, keep only 31 bits
 
-        return str_pad((string) ($value % 10** $this->digits), $this->digits, '0', STR_PAD_LEFT);
+        return str_pad((string)($value % 10 ** $this->digits), $this->digits, '0', STR_PAD_LEFT);
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:82]: good spot on the missing space but I find a space after the type coercion much more readable

File: lib/TwoFactorAuth.php [hunk 11]
```diff
@@ -157,29 +158,23 @@ public function getQRText(string $label, string $secret): string
         return 'otpauth://totp/' . rawurlencode($label)
             . '?secret=' . rawurlencode($secret)
             . '&issuer=' . rawurlencode((string)$this->issuer)
-            . '&period=' . intval($this->period)
+            . '&period=' . $this->period
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:161]: good spot on the unnecessary `intval()`'s but should this be a `http_build_query()`?

File: lib/TwoFactorAuth.php [hunk 12]
```diff
@@ -157,29 +158,23 @@ public function getQRText(string $label, string $secret): string
...
-     * @throws TwoFactorAuthException
-     */
     public function getQrCodeProvider(): IQRCodeProvider
     {
         // Set default QR Code provider if none was specified
-        if (null === $this->qrcodeprovider) {
-            return $this->qrcodeprovider = new QRServerProvider();
-        }
-        return $this->qrcodeprovider;
+        return $this->qrcodeprovider ?? ($this->qrcodeprovider = new QRServerProvider());
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:169]: isn't this what `??=` is for? (Null Coalescing Assignment)

File: lib/TwoFactorAuth.php [hunk 13]
```diff
@@ -194,16 +189,10 @@ public function getRngProvider(): IRNGProvider
...
-     * @throws TwoFactorAuthException
-     */
     public function getTimeProvider(): ITimeProvider
     {
         // Set default time provider if none was specified
-        if (null === $this->timeprovider) {
-            return $this->timeprovider = new LocalMachineTimeProvider();
-        }
-        return $this->timeprovider;
+        return $this->timeprovider ?? ($this->timeprovider = new LocalMachineTimeProvider());
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:195]: isn't this what ??= is for? (Null Coalescing Assignment)

File: lib/TwoFactorAuth.php [hunk 14]
```diff
@@ -218,7 +207,8 @@ private function codeEquals(string $safe, string $user): bool
         // we don't leak information about the difference of the two strings.
         if (strlen($safe) === strlen($user)) {
             $result = 0;
-            for ($i = 0; $i < strlen($safe); $i++) {
+            $strlen = strlen($safe);
+            for ($i = 0; $i < $strlen; $i++) {
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:211]: I think separating the limit out to a variable makes this less readable

File: lib/TwoFactorAuth.php [hunk 15]
```diff
@@ -228,21 +218,21 @@ private function codeEquals(string $safe, string $user): bool
...
     private function getTime(?int $time = null): int
     {
-        return ($time === null) ? $this->getTimeProvider()->getTime() : $time;
+        return $time ?? $this->getTimeProvider()->getTime();
     }
 
     private function getTimeSlice(?int $time = null, int $offset = 0): int
     {
-        return (int) floor($time / $this->period) + ($offset * $this->period);
+        return (int)floor($time / $this->period) + ($offset * $this->period);
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:226]: I find a space after the type coercion much more readable

File: lib/TwoFactorAuth.php [hunk 16]
```diff
@@ -228,21 +218,21 @@ private function codeEquals(string $safe, string $user): bool
...
 
     private function base32Decode(string $value): string
     {
-        if (strlen($value) == 0) {
+        if ($value === '') {
             return '';
         }
 
-        if (preg_match('/[^' . preg_quote(self::$_base32dict) . ']/', $value) !== 0) {
+        if (preg_match('/[^' . preg_quote(self::$_base32dict, '/') . ']/', $value) !== 0) {
```
Comment (willpower232) [inline lib/TwoFactorAuth.php:235]: I don't think this change accomplishes anything unless you know PHP is planning to change the default delimiter?

File: tests/Providers/Rng/TestRNGProvider.php [hunk 17]
```diff
@@ -8,7 +8,7 @@
 
 class TestRNGProvider implements IRNGProvider
 {
-    public function __construct(private bool $isSecure = false)
+    public function __construct(private readonly bool $isSecure = false)
```
Comment (willpower232) [inline tests/Providers/Rng/TestRNGProvider.php:11]: potentially unnecessary readonly as previously discussed

File: lib/Providers/Rng/OpenSSLRNGProvider.php [hunk 18]
```diff
@@ -6,7 +6,7 @@
 
 class OpenSSLRNGProvider implements IRNGProvider
 {
-    public function __construct(private bool $requirestrong = true)
+    public function __construct(private readonly bool $requirestrong = true)
```
Comment (willpower232) [inline lib/Providers/Rng/OpenSSLRNGProvider.php:9]: potentially unnecessary readonly as previously discussed

File: lib/Providers/Rng/HashRNGProvider.php [hunk 19]
```diff
@@ -8,7 +8,7 @@
 
 class HashRNGProvider implements IRNGProvider
 {
-    public function __construct(private string $algorithm = 'sha256')
+    public function __construct(private readonly string $algorithm = 'sha256')
```
Comment (willpower232) [inline lib/Providers/Rng/HashRNGProvider.php:11]: potentially unnecessary readonly as previously discussed

**Review by willpower232 — DISMISSED** [review]
Thanks for your work, always good to have fresh eyes on something.

I've left a few points for discussion (mostly duplicated) so I look forward to your replies.

**Comment by RobThree** [comment]
Generally such 'huge' PR's aren't very helpful, this is so much that it's hard to see what has changed and what hasn't. Next time, please do one thing, make it into a PR and then make a new PR for the next thing.

I agree with most, if not all, of @willpower232's 'criticisms' above. I've just returned home from a weekend away; I'll go over it myself soon.

That's not to say the PR and your work isn't appreciated, just to be clear!

File: lib/Providers/Qr/BaconQrCodeProvider.php (hunk 4 above)
Comment (xHeaven) [reply lib/Providers/Qr/BaconQrCodeProvider.php:28]: It wasn't me who did this, though, it was the PHP-CS-Fixer.

File: lib/Providers/Qr/BaconQrCodeProvider.php (hunk 6 above)
Comment (xHeaven) [reply lib/Providers/Qr/BaconQrCodeProvider.php:81]: `array()` vs `[]` is configured in the PHP-CS-Fixer rules. I can change the rules, but it will modify quite some files.

File: lib/Providers/Time/NTPTimeProvider.php (hunk 7 above)
Comment (xHeaven) [reply lib/Providers/Time/NTPTimeProvider.php:50]: Another PHP-CS-Fixer rule.

File: lib/TwoFactorAuth.php (hunk 8 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:37]: This is a PHP-CS-Fixer rule as well.

File: lib/TwoFactorAuth.php (hunk 10 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:82]: This was also PHP-CS-Fixer.

File: lib/TwoFactorAuth.php (hunk 11 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:161]: Yeah, I think it can be converted into a `http_build_query`.

File: lib/TwoFactorAuth.php (hunk 14 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:211]: I beg to differ, also recounting the strlen in every loop is resource intensive, depending on the input.

File: lib/TwoFactorAuth.php (hunk 15 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:226]: Another PHP-CS-Fixer rule.

**Comment by xHeaven** [comment]
Appreciate the replies and reviews. I've answered some of them, I'll answer all of them a bit later today.

File: lib/Providers/Qr/BaconQrCodeProvider.php (hunk 3 above)
Comment (xHeaven) [reply lib/Providers/Qr/BaconQrCodeProvider.php:25]: `borderWidth` is the only variable here that doesnt get modified anywhere in the code. The rest of the properties are overwritten in the constructor by various methods.

File: lib/Providers/Time/HttpTimeProvider.php (hunk 2 above)
Comment (xHeaven) [reply lib/Providers/Time/HttpTimeProvider.php]: `$options` is now defined by constructor property promotion in https://github.com/RobThree/TwoFactorAuth/pull/108/commits/c2183e16b372bf7c3303d437f00a326ca964771e.
As for the formatting, that's PHP-CS-Fixer's doing.

File: lib/TwoFactorAuth.php (hunk 16 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:235]: That's right, since `$_base32dict` doesn't contain any `/` character, it's a bit meaningless. I also don't think PHP is going to change the default delimiter anytime soon. Having it, just in case though, can't hurt, can it?

File: tests/Providers/Rng/TestRNGProvider.php (hunk 17 above)
Comment (xHeaven) [reply tests/Providers/Rng/TestRNGProvider.php:11]: Adding the readonly modifier is useful so the developers instantly know that it's only set in the constructor and is not modified later on anywhere, not even in its own class.
In this case, I kind of understand it as there is really not much happening here, however, I put it there for consistency's sake. In my opinion, if a variable is only set once and it's never going to get modified later on, it should be prefixed with `readonly`. Not prefixing it could mean that you might be able to modify its value somewhere in its own class.

File: lib/Providers/Rng/OpenSSLRNGProvider.php (hunk 18 above)
Comment (xHeaven) [reply lib/Providers/Rng/OpenSSLRNGProvider.php:9]: See: https://github.com/RobThree/TwoFactorAuth/pull/108#discussion_r1208601635

File: lib/Providers/Rng/HashRNGProvider.php (hunk 19 above)
Comment (xHeaven) [reply lib/Providers/Rng/HashRNGProvider.php:11]: See: https://github.com/RobThree/TwoFactorAuth/pull/108#discussion_r1208601635

File: lib/TwoFactorAuth.php (hunk 9 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php:57]: This is also PHP-CS-Fixer's doing. It's in the rules to format it like such.

File: lib/TwoFactorAuth.php (hunk 13 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php]: Good call. Fixed in https://github.com/RobThree/TwoFactorAuth/pull/108/commits/9bc454f42501a89ce98860e1e9b7b774e555bf38.

File: lib/TwoFactorAuth.php (hunk 12 above)
Comment (xHeaven) [reply lib/TwoFactorAuth.php]: Good call. Fixed in https://github.com/RobThree/TwoFactorAuth/pull/108/commits/9bc454f42501a89ce98860e1e9b7b774e555bf38.

**Comment by xHeaven** [comment]
I've answered all of your reviews and marked most of them as resolved. If you've got a spare 5 minutes sometimes, could you take a look at it? @willpower232 Thanks!

File: lib/Providers/Qr/BaconQrCodeProvider.php (hunk 5 above)
Comment (willpower232) [reply lib/Providers/Qr/BaconQrCodeProvider.php:54]: aha I'd vote for camelCase everywhere in that case!

**Comment by willpower232** [comment]
I wasn't aware the PHP CS Fixer config was so extra compared to just the bare standard, I usually just choose to run `phpcs --standard=PSR12 app/` and fix as necessary :sweat_smile: 

I think this PR size is fine since its all overlapping stuff and the codebase has aged differently in different places with the evolution of PHP since it was first done. I don't think another PR of this size would exist in this repo for some time!

Thanks for all the explanations and indulging my many duplicates, I think I'm happy once the camelCase has happened.

File: lib/Providers/Qr/BaconQrCodeProvider.php (hunk 5 above)
Comment (xHeaven) [reply lib/Providers/Qr/BaconQrCodeProvider.php]: Here you go, fixed in https://github.com/RobThree/TwoFactorAuth/pull/108/commits/e554a9b5e482d9e8e9b72353d8e23501e663607c :)

**Comment by xHeaven** [comment]
Yeah, the Fixer config has quite a few different rules other than just PSR12, haha.
I've committed the camelCase changes as well in https://github.com/RobThree/TwoFactorAuth/pull/108/commits/e554a9b5e482d9e8e9b72353d8e23501e663607c.

**Review by willpower232 — COMMENTED** [review]
happy with this if @RobThree is

**Comment by RobThree** [comment]
> happy with this if @RobThree is

I am. Not sure I agree on the whitespace, but that's not worth a discussion.

Thanks @xHeaven! I'll merge.
